alembic/script/revision.py

```diff
@@ -886,6 +895,122 @@ def _iterate_related_revisions(self, fn, targets, map_, check=False):
                         )
                     )
 
+    def topological_sort(self, revisions, reverse=False):
+        """Yield revision ids of a collection of Revision objects in
+        topological sorted order (i.e. revisions always come after their
+        down_revisions and dependencies)."""
+        allitems = [d.revision for d in revisions]
+        edges = [
+            (rev, child.revision)
+            for child in revisions
+            if child.down_revision is not None
+            for rev in util.to_tuple(child.down_revision)
+        ] + [
+            (parent, child.revision)
+            for child in revisions
+            if child.dependencies is not None
+            for parent in util.to_tuple(child.dependencies)
+        ]
+        return sqlautil.topological.sort(edges, allitems)
+
+    def walk_down(self, start, steps, label):
+        """ Walk down the tree along a single path. """
+        assert steps <= 0
+        if isinstance(start, compat.string_types):
+            start = self.get_revision(start)
+        if steps == 0:
+            return start
+        if start is None:
+            raise ValueError("Can't walk back past base.")
+        children = self.get_revisions(start.down_revision)
+        if len(children) == 0:
+            raise util.CommandError(
```

Comment:
**mike bayer** (zzzeek) wrote:

still supporting python 2.7:

        if len(children) == 0:
            raise util.CommandError(
                "Relative revision %(label)s didn't produce "
                "%(abslabel)s  migrations" % {
                    "label": label,
                    "abslabel": abs(int(label))
                }
            )

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530

Reply:
**Simon Bowly** (simonbowly) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2530